lib/asm-parser-ewavr.js

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2019, Ethan Slattery
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+const AsmParserBase = require('./asm-parser'),
+    utils = require('./utils'),
+    AsmRegex = require('./asmregex').AsmRegex;
+
+class AsmParser extends AsmParserBase {
+    processAsm(asm, filters) {
+        if (filters.binary) return this.processBinaryAsm(asm, filters);
+
+        if (filters.commentOnly) {
+            // Remove any block comments that start and end on a line if we're removing comment-only lines.
+            const blockComments = /^\s*\/\*(\*(?!\/)|[^*])*\*\/\s*\r?\n/mg;
+            asm = asm.replace(blockComments, "");
+        }
+
+        const result = [];
+        let asmLines = utils.splitLines(asm);
+        if (filters.preProcessLines !== undefined) {
+            asmLines = filters.preProcessLines(asmLines);
+        }
+
+        const labelsUsed = this.findUsedLabels(asmLines, filters.directives);
+        let prevLabel = "";
+
+        const commentOnly = /^\s*(((#|@|;|\/\/).*)|(\/\*.*\*\/))$/;
+        const commentOnlyNvcc = /^\s*(((#|;|\/\/).*)|(\/\*.*\*\/))$/;
+        const sourceTag = /^\/\/\s*([0-9]+)\s(?!bytes).*/;
+        const source6502Dbg = /^\s*\.dbg\s+line,\s*"([^"]+)",\s*(\d+)/;
+        const source6502DbgEnd = /^\s*\.dbg\s+line[^,]/;
+        const sourceStab = /^\s*\.stabn\s+(\d+),0,(\d+),.*/;
+        const stdInLooking = /<stdin>|^-$|example\.[^/]+$|<source>/;
+        const endBlock = /\.(cfi_endproc|data|text|section)/;
```

Comment:
There is a lot of code duplication between this class and the base class:
https://github.com/mattgodbolt/compiler-explorer/blob/9f055b12948e85e6d43235c2e2d78faf27fe2439/lib/asm-parser.js#L221-L228

You only changed the `sourceTag` regex. Maybe we can add these variables as a member of the AsmParser in the constructor:
https://github.com/mattgodbolt/compiler-explorer/blob/9f055b12948e85e6d43235c2e2d78faf27fe2439/lib/asm-parser.js#L29-L52

And you can override `sourceTag` in the constructor of your class.

Reply:
Agreed! Thanks @csordasmarton for your extra comments, I agree completely:)

Reply:
I made most of these changes. But this last one I am unsure about. It looks like the return type of `processAsm()` has changed since I last rebased, so I had to grab the latest version to make it work. Since the newer version parses out the file number from `.loc` I have more changes (IAR doesn't have the file number in each source tag).

- sourceTag regex is different
- definition of `handleSource()` is different, to force file to null

I also noticed that the vast majority of `processAsm()` is useless for IAR, since there will be no stab, nvcc, 6502, etc. If `sourceTag` could be set in a constructor and I just override it how does that now solve the `handleSource()` function being different?

I could just keep overloading the `processAsm()` method, but cut out all the extras, making it _much much_ smaller.